### ISSUE TITLE
Accessing input training data from `Dataset` object? 

### ISSUE THREAD

**Issue by IanQS** [Microsoft/LightGBM#1690]
## Environment info

Operating System: Ubuntu 16.04

CPU: Yes (not GPU if that's what this means?)

C++/Python/R version: Python

In [advanced example](https://github.com/Microsoft/LightGBM/blob/master/examples/python-guide/advanced_example.py) we got the labels from the training data but I want to get the features themselves. How would I do so? 

I need to implement a custom loss function and I need acces to the `X's` for a custom regression task. Specifically, I want to do a regression, and I'd like to include a regularizer. The regularizer would look like `np.inner((label - X[-1], (predicted - X[-1])`) having the same sign

**Comment by StrikerRUS** [comment]
Hi @IanQS !

You can access raw data via the `data` field. It's documented in [**FAQ Python-package Solution 2**](https://lightgbm.readthedocs.io/en/latest/FAQ.html#python-package).

Don't forget to set `free_raw_data=False` while constructing Dataset. And in custom loss you'll be able to get X's (assuming you have the following function signature: `def custom_loss(preds, train_data):`):
```
X = train_data.data
```

**Comment by IanQS** [comment]
@StrikerRUS 

Thank you so much! Is there a way to get the data that lightgbm has processed? If I understand correctly it does some preprocessing on its end to reduce dimensionality / format the data?

**Comment by StrikerRUS** [comment]
It's `Dataset` object itself:
> The Dataset object in LightGBM is very memory-efficient, due to it only need to save discrete bins.

Unfortunately, you can only save it as binary file:
https://lightgbm.readthedocs.io/en/latest/Python-API.html#lightgbm.Dataset.save_binary
